lib/src/git.rs

```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
-                .expect("Expect reference name to be valid UTF-8");
-            let expected_remote_location = *qualified_remote_refs_expected_locations
-                .get(dst_refname)
-                .expect("Push is trying to move a ref it wasn't asked to move");
-            let expected_location_oid = expected_remote_location
-                .map(|id| git2::Oid::from_bytes(id.as_bytes()).unwrap())
-                .unwrap_or(git2::Oid::zero());
-            // `update.src()` is the current actual position of the branch
-            // `dst_refname` on the remote. `update.dst()` is the position
-            // we are trying to push the branch to (AKA our local branch
-            // location). 0000000000 is a valid value for either, indicating
-            // branch creation or deletion.
-
-            // Short-term TODO: Replace this with actual rules of when to permit the push
-            tracing::info!(
-                "Forcing {dst_refname} to {dst}. It was at {src} at the server. We expected it to \
-                 be at {expected_location_oid}.",
-                src = update.src(),
-                dst = update.dst()
-            );
-        }
-        Ok(())
-    });
-    callbacks.push_update_reference(|refname, status| {
-        // The status is Some if the ref update was rejected
-        if status.is_none() {
-            remaining_remote_refs.remove(refname);
-        }
-        Ok(())
-    });
-    push_options.remote_callbacks(callbacks);
-    remote
-        .push(refspecs, Some(&mut push_options))
-        .map_err(|err| match (err.class(), err.code()) {
-            (git2::ErrorClass::Reference, git2::ErrorCode::NotFastForward) => {
-                GitPushError::NotFastForward
+    let mut failed_push_negotiations = vec![];
+    let push_result = {
+        let mut push_options = git2::PushOptions::new();
+        let mut proxy_options = git2::ProxyOptions::new();
+        proxy_options.auto();
+        push_options.proxy_options(proxy_options);
+        let mut callbacks = callbacks.into_git();
+        callbacks.push_negotiation(|updates| {
+            for update in updates {
+                let dst_refname = update
+                    .dst_refname()
+                    .expect("Expect reference name to be valid UTF-8");
+                let expected_remote_location = *qualified_remote_refs_expected_locations
+                    .get(dst_refname)
+                    .expect("Push is trying to move a ref it wasn't asked to move");
+                let oid_to_maybe_commitid =
+                    |oid: git2::Oid| (!oid.is_zero()).then(|| CommitId::from_bytes(oid.as_bytes()));
+                let actual_remote_location = oid_to_maybe_commitid(update.src());
+                let local_location = oid_to_maybe_commitid(update.dst());
+
+                let location_to_string = |loc: Option<&CommitId>| {
+                    loc.map(|id| id.hex()).unwrap_or("(missing)".to_string())
+                };
```

Comment:
nit: just use debug format?

I'm not strongly against adding long tracing message, but it doesn't have to be overly user-friendly. I think the debug formatting of `Option<&CommidId>` is pretty okay.

Reply:
You're probably right.

I was trying to come up with something that distinguishes "no expectation" from "expected to not exist", as you pointed out in one of your old comments. On second thought, "(missing)" is not much better than None in this regard. If I don't think of something better, I'll probably just switch to debug formatting.